Seperating the building of imageName, appVersionStr and packagesParameter

brujoand opened the pull request:
The rationale for this is that when baking docker images in Rosco the image names ends up quite convoluted, when the desired result is simply having the ami-name as image name, maybe with the possibility to add tags as appVersionStr. By splitting up the building of imageName, appVersionStr and packagesParameter I hope to make it easier to enable provider specific naming conventions. I looked at the possibility of just using 'extended_attributes', but to me this seems like something which will be needed by others also, so I went for the long road instead. 

This is my second stab at groovy so feel free to nitpick. Also, I'm on vacation, so I apologise if response isn't immediate.

lwander commented:
This might be to early to ask - but how do you plan on including the "organization" in the docker image name? Usually docker images follow the convention

```
<registry.domain>/<organization>/<image>:<tag>
```

Since "organization" is now a first class object in the [Docker Trigger configuration](https://github.com/spinnaker/deck/pull/2306#issuecomment-223418275), it would make sense to surface that somehow here.

brujoand commented:
Good question. My initial thought would be to expect an ami_name on the form of "myregistry.io/organization/repository" and then override buildImageName by parsing the ami_name. I guess this would be the easiest route, but increases probability of error. It also has to work with subsets of that ami_name etc. 
But this is very much colored by the fact that I hadn't noticed the change in the docker trigger configuration. What's your thoughts on this?

lwander commented:
I think ideally we would surface Registry (a pre-configured docker account in clouddriver), Organization and Image all as parameters to the build, which will likely be populated by dropdown. Tag will then be a function of package details/commit hash/date/etc... to ensure it's unique. This way it remains consistent with other references to Docker in Spinnaker, and will be easier to integrate with Kubernetes.

brujoand commented:
Sounds like a good approach to me. Do you see these new arguments passed to the build as explicit docker parameters or added to the extended attributes? 

While on the topic; the handling of docker tags has been puzzling me a bit. Some of those I've talked to wants to be able to use multiple tags pr image. But I guess that could be fitted in the extended attributes. Have you given that part any thought?

lwander commented:
I think where the parameters go is best left to @duftler (he's OOO so he might not respond for a bit).

When you mean multiple tags per image, do you mean the same (identical contents) image is tagged with several tags? I think that's a good practice in general and we should enable that.

brujoand commented:
Cool :thumbsup:
Yeah, same contents with multiple tags would be nice.

brujoand commented:
Merged with master, fixed conflict, squashed commits and pushed with force.

duftler commented on rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/CloudProviderBakeHandler.groovy:
```diff
@@ -140,7 +141,21 @@ abstract class CloudProviderBakeHandler {
+    }
+
+    if (!osPackages.empty) {
+     appVersionStr = imageNameFactory.buildAppVersionStr(bakeRequest, osPackages.first())
```
Any reason not to make `imageNameFactory.buildAppVersionStr()` take the full list of packages, instead of extracting the first package here at the call-site, so it is more consistent with its siblings?

brujoand replied:
Initially it made sense since I'm basically just refactoring out the existing behaviour which only uses the first package. But looking at it again I think it's a good idea to send inn the entire list as it might be useful for someone else, and it puts the logic of making this choice where it belongs; within the method which can be overridden.

duftler commented on rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/util/PackageNameConverter.groovy:
```diff
@@ -66,6 +70,7 @@ class PackageNameConverter {
   // For example: nflx-djangobase-enhanced-0.1-h12.170cdbd.all
   public static OsPackageName parseRpmPackageName(String fullyQualifiedPackageName) {
     OsPackageName osPackageName = new OsPackageName()
+    String versionDelimiter = BakeRequest.PackageType.RPM.getVersionDelimiter()
```
Noticed that you refer to `versionDelimiter` up above against `DEB`, but used `getVersionDelimiter()` here for `RPM`.

brujoand replied:
ah, using the getter for both now. I'm a bit surprised that the former worked since it's marked private. I guess it's a bug of sorts. https://issues.apache.org/jira/browse/GROOVY-3010

duftler commented on rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/util/PackageNameConverter.groovy:
```diff
@@ -82,6 +87,9 @@ class PackageNameConverter {
           release = parts.pop()
           version = parts.pop()
           name = parts.join("-")
+          qualifiedName = "${name}${versionDelimiter}${version}-${release}"
```
Since `qualifiedName` is composed the same way for both package types, maybe just make it a method on `OsPackageName`? It can still return `null` if `release` is missing.

brujoand replied:
We would need the versionDelimiter in the method. Having it as a part of the OsPackageName seems random. But passing it as a parameter makes more sense in that it keeps the OsPackageName agnostic to what PackageType it's derived from. I'll try that.

duftler commented on rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/util/ImageNameFactory.groovy:
```diff
@@ -16,27 +16,73 @@
+   * Returns packagesParameter; the updated list of packages to be used for overriding the passed package list
+   */
+  def buildPackagesParameter(List<PackageNameConverter.OsPackageName> osPackageNames) {
+    if (osPackageNames.empty) return ""
```
You should verify this, but I think is unnecessary.  `.collect` should return an empty list, which should produce an empty string when `.join` is applied to it.

duftler commented on rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/CloudProviderBakeHandler.groovy:
```diff
@@ -140,7 +141,21 @@ abstract class CloudProviderBakeHandler {
     BakeOptions.Selected selectedOptions = new BakeOptions.Selected(baseImage: findBaseImage(bakeRequest))
-    def (imageName, appVersionStr, packagesParameter) = imageNameFactory.deriveImageNameAndAppVersion(bakeRequest, selectedOptions)
+
+    List<String> packageNameList = bakeRequest?.package_name?.tokenize(" ")
```
Since `packageNameList` can end up `null`, and the `.collect` down below will return an empty list when applied to it, the `osPackageNames.first()` just inside of `buildImageName()` will likely throw an exception when `package_name` is `null` or an empty string.

duftler replied:
I didn't get to the unit tests yet, but we should probably have explicit tests for a `null` or empty string `package_name`.

duftler replied:
`bakeRequest` can't really be `null` here, so we don't need the null safe navigation between it and `package_name`.

brujoand replied:
Good point. I'll add a test for empty/null packageNameList and remove the guard on bakeRequest.

duftler commented on rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/google/GCEBakeHandlerSpec.groovy:
```diff
@@ -159,6 +160,9 @@ class GCEBakeHandlerSpec extends Specification {
                                         base_os: "ubuntu",
                                         cloud_provider_type: BakeRequest.CloudProviderType.gce)
       def targetImageName = "kato-x8664-timestamp-ubuntu"
+      def osPackages = bakeRequest.package_name.split(" ").collect { packageName ->
```
This pattern appears enough that it might make sense to extract it into a utility function for use in all the unit tests.

duftler commented on rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/util/ImageNameFactory.groovy:
```diff
@@ -16,27 +16,73 @@
+   */
+  def buildImageName(BakeRequest bakeRequest, List<PackageNameConverter.OsPackageName> osPackageNames) {
+    String timestamp = clock.millis()
+    def emptyBasePackage = new PackageNameConverter.OsPackageName()
```
This feels a little like abusing the `OsPackageName` model, especially since the `buildOsPackageName()` factory method throws an exception if it's not `DEB` or `RPM`. Maybe add an explicit type there for this use case?
If you decide you really want to keep it this way, you should be able to use: `def emptyBasePackage = new PackageNameConverter.OsPackageName(name: '', arch: 'all')`

brujoand replied:
Yeah, I agree. I aimed at reducing ``` either :? or ``` for readability but it ended up more cryptic I think. Simplifying this instead.

duftler commented on rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/util/ImageNameFactory.groovy:
```diff
@@ -16,27 +16,73 @@
+    String release = bakeRequest.ami_suffix ?: timestamp
+    String arch = basePackage.arch ?: "all" // TODO This is defaulted to 'all' since all tests assume this. Skip?
+
+    [ baseName, arch, release, bakeRequest.base_os].findAll{it != "" }.join("-")
```
Removing the leading space before `baseName`, or adding a trailing space after `bakeRequest.base_os`, would increase readability. Same for the space surrounding `it != ""`.

duftler replied:
Since the test you're using here is `it != ""`, this leaves us open to concatenating `null` when one of the items in the list is explicitly `null`. While this looks impossible with the current path to this line, I think it would be more future-proof if the test were actually just `it`. Then the result would be false for both an empty string and a `null`.

brujoand replied:
Oh good one. Checking just ```it``` looks much better.

duftler commented on rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/util/ImageNameFactory.groovy:
```diff
@@ -16,27 +16,73 @@
+    PackageNameConverter.OsPackageName basePackage = osPackageNames.first() ?: emptyBasePackage
+
+    String baseName = bakeRequest.ami_name ?: basePackage.name
+    String release = bakeRequest.ami_suffix ?: timestamp
```
Might be more readable to have the order here match the order they are used in composing the image name down below.

brujoand replied:
👍

duftler commented on rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/util/ImageNameFactorySpec.groovy:
```diff
@@ -142,15 +155,20 @@ class DefaultImageNameFactorySpec extends Specification {
     when:
-      def (imageName, appVersionStr, packagesParameter) = imageNameFactory.deriveImageNameAndAppVersion(bakeRequest, selectedOptions)
+      def imageName = imageNameFactory.buildImageName(bakeRequest, osPackages)
+      def appVersionStr= imageNameFactory.buildAppVersionStr(bakeRequest, osPackages.head())
```
Using `head()` in some spots and `first()` in others may result in future readers searching around to see if they behave the same. We should probably just stick to one (I prefer `first()` but don't feel that strongly about it).

duftler replied:
Missing a space before the =.

duftler commented on rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/CloudProviderBakeHandler.groovy:
```diff
@@ -140,7 +141,21 @@ abstract class CloudProviderBakeHandler {
+    List<String> packageNameList = bakeRequest?.package_name?.tokenize(" ")
+    def appVersionStr
+
+    def osPackages = packageNameList.collect { name ->
```
Any reason not to call this `osPackageNames`, like the parameter names in the methods it is passed to?

brujoand replied:
No, it makes sense to use ```osPackageNames``` here also

duftler commented on rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/aws/AWSBakeHandlerSpec.groovy:
```diff
@@ -628,7 +672,7 @@ class AWSBakeHandlerSpec extends Specification {
         repository: DEBIAN_REPOSITORY,
         package_type: BakeRequest.PackageType.DEB.packageType,
-        packages: fullyQualifiedPackageName,
+        packages: unqualifiedPackageName,
```
What is the reason that this particular `package_name` parameter now results in an unqualified package name being returned?
It looks like it's just because of this change: `1 * imageNameFactoryMock.buildPackagesParameter(osPackages) >> unqualifiedPackageName`
But why is that change made to the particular test case?

brujoand replied:
Oh wow. Thanks for pointing this out. I just read up on the spock framework and the right shift operator. I've misunderstood this: ``` 1 * imageNameFactoryMock.buildPackagesParameter(osPackages) >> unqualifiedPackageName ``` to mean ``` verify that this method should be called once and that it returns unqualifiedPackageName ``` when the return actually tells the mock what it should return. Sorry about that, I'll take a fresh look at the tests with this in mind.

duftler commented on rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/google/GCEBakeHandlerSpec.groovy:
```diff
@@ -503,8 +545,9 @@ class GCEBakeHandlerSpec extends Specification {
     then:
-    1 * imageNameFactoryMock.deriveImageNameAndAppVersion(bakeRequest, _) >> [targetImageName, null, PACKAGES_NAME]
-      IllegalArgumentException e = thrown()
+    1 * imageNameFactoryMock.buildImageName(bakeRequest, osPackages) >> targetImageName
```
These 3 lines should be indented 2 more spaces.

duftler commented on rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/util/ImageNameFactorySpec.groovy:
```diff
@@ -124,13 +133,17 @@ class DefaultImageNameFactorySpec extends Specification {
     when:
-      def (imageName, appVersionStr, packagesParameter) = imageNameFactory.deriveImageNameAndAppVersion(bakeRequest, selectedOptions)
+      def imageName = imageNameFactory.buildImageName(bakeRequest, [osPackage])
+      print(imageName)
```
Extraneous?

duftler commented on rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/util/ImageNameFactorySpec.groovy:
```diff
@@ -279,13 +318,17 @@ class DefaultImageNameFactorySpec extends Specification {
     when:
-      def (imageName, appVersionStr, packagesParameter) = imageNameFactory.deriveImageNameAndAppVersion(bakeRequest, selectedOptions)
+      def imageName = imageNameFactory.buildImageName(bakeRequest, osPackages)
+      def appVersionStr= imageNameFactory.buildAppVersionStr(bakeRequest, osPackages.head())
```
Missing a space before the `=`.

duftler commented on rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/util/ImageNameFactorySpec.groovy:
```diff
@@ -259,15 +294,19 @@ class DefaultImageNameFactorySpec extends Specification {
     when:
-      def (imageName, appVersionStr, packagesParameter) = imageNameFactory.deriveImageNameAndAppVersion(bakeRequest, selectedOptions)
+      def imageName = imageNameFactory.buildImageName(bakeRequest, osPackages)
+      def appVersionStr= imageNameFactory.buildAppVersionStr(bakeRequest, osPackages.head())
```
Missing a space before the =.

duftler commented on rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/util/ImageNameFactorySpec.groovy:
```diff
@@ -240,14 +272,17 @@ class DefaultImageNameFactorySpec extends Specification {
     when:
-      def (imageName, appVersionStr, packagesParameter) = imageNameFactory.deriveImageNameAndAppVersion(bakeRequest, selectedOptions)
+      def imageName = imageNameFactory.buildImageName(bakeRequest, [osPackage])
+      def appVersionStr= imageNameFactory.buildAppVersionStr(bakeRequest, osPackage)
```
Missing a space before the =.

duftler commented on rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/util/ImageNameFactorySpec.groovy:
```diff
@@ -220,15 +249,18 @@ class DefaultImageNameFactorySpec extends Specification {
     when:
-      def (imageName, appVersionStr, packagesParameter) = imageNameFactory.deriveImageNameAndAppVersion(bakeRequest, selectedOptions)
+      def imageName = imageNameFactory.buildImageName(bakeRequest, [osPackage])
+      def appVersionStr= imageNameFactory.buildAppVersionStr(bakeRequest, osPackage)
```
Missing a space before the =.

duftler commented on rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/util/ImageNameFactorySpec.groovy:
```diff
@@ -202,13 +228,16 @@ class DefaultImageNameFactorySpec extends Specification {
     when:
-      def (imageName, appVersionStr, packagesParameter) = imageNameFactory.deriveImageNameAndAppVersion(bakeRequest, selectedOptions)
+      def imageName = imageNameFactory.buildImageName(bakeRequest, [osPackage])
+      def appVersionStr= imageNameFactory.buildAppVersionStr(bakeRequest, osPackage)
```
Missing a space before the =.

duftler commented on rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/util/ImageNameFactorySpec.groovy:
```diff
@@ -182,15 +204,19 @@ class DefaultImageNameFactorySpec extends Specification {
     when:
-      def (imageName, appVersionStr, packagesParameter) = imageNameFactory.deriveImageNameAndAppVersion(bakeRequest, selectedOptions)
+      def imageName = imageNameFactory.buildImageName(bakeRequest, osPackages)
+      def appVersionStr= imageNameFactory.buildAppVersionStr(bakeRequest, osPackages.head())
```
Missing a space before the =.

duftler commented on rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/util/ImageNameFactorySpec.groovy:
```diff
@@ -162,15 +180,19 @@ class DefaultImageNameFactorySpec extends Specification {
     when:
-      def (imageName, appVersionStr, packagesParameter) = imageNameFactory.deriveImageNameAndAppVersion(bakeRequest, selectedOptions)
+      def imageName = imageNameFactory.buildImageName(bakeRequest, osPackages)
+      def appVersionStr= imageNameFactory.buildAppVersionStr(bakeRequest, osPackages.head())
```
Missing a space before the =.

duftler commented on rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/util/ImageNameFactorySpec.groovy:
```diff
@@ -104,15 +111,17 @@ class DefaultImageNameFactorySpec extends Specification {
     when:
-      def (imageName, appVersionStr, packagesParameter) = imageNameFactory.deriveImageNameAndAppVersion(bakeRequest, selectedOptions)
+    def imageName = imageNameFactory.buildImageName(bakeRequest, [osPackage])
+    def appVersionStr= imageNameFactory.buildAppVersionStr(bakeRequest, osPackage)
```
Missing a space before the =.

duftler commented on rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/util/ImageNameFactorySpec.groovy:
```diff
@@ -85,14 +90,16 @@ class DefaultImageNameFactorySpec extends Specification {
     when:
-     def (imageName, appVersionStr, packagesParameter) = imageNameFactory.deriveImageNameAndAppVersion(bakeRequest, selectedOptions)
+      def imageName = imageNameFactory.buildImageName(bakeRequest, [osPackage])
+      def appVersionStr= imageNameFactory.buildAppVersionStr(bakeRequest, osPackage)
```
Missing a space before the =.

duftler commented on rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/util/ImageNameFactorySpec.groovy:
```diff
@@ -66,14 +69,16 @@ class DefaultImageNameFactorySpec extends Specification {
     when:
-      def (imageName, appVersionStr, packagesParameter) = imageNameFactory.deriveImageNameAndAppVersion(bakeRequest, selectedOptions)
+      def imageName = imageNameFactory.buildImageName(bakeRequest, [osPackage])
+      def appVersionStr= imageNameFactory.buildAppVersionStr(bakeRequest, osPackage)
```
Missing a space before the =.

duftler commented on rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/util/ImageNameFactorySpec.groovy:
```diff
@@ -47,14 +48,16 @@ class DefaultImageNameFactorySpec extends Specification {
     when:
-      def (imageName, appVersionStr, packagesParameter) = imageNameFactory.deriveImageNameAndAppVersion(bakeRequest, selectedOptions)
+      def imageName = imageNameFactory.buildImageName(bakeRequest, [osPackage])
+      def appVersionStr= imageNameFactory.buildAppVersionStr(bakeRequest, osPackage)
```
Missing a space before the =.

duftler commented on rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/util/ImageNameFactorySpec.groovy:
```diff
@@ -16,26 +16,27 @@
     when:
-      def (imageName, appVersionStr, packagesParameter) = imageNameFactory.deriveImageNameAndAppVersion(bakeRequest, selectedOptions)
+      def imageName = imageNameFactory.buildImageName(bakeRequest, [osPackage])
+      def appVersionStr= imageNameFactory.buildAppVersionStr(bakeRequest, osPackage)
```
Missing a space before the =.

duftler commented on rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/util/PackageNameConverterSpec.groovy:
```diff
@@ -99,7 +106,7 @@ class PackageNameConverterSpec extends Specification {
 
     then:
-      parsedDebPackageName == parsedRpmPackageName
+      parsedDebPackageName != parsedRpmPackageName
```
Seems like we could compare each of the parts here to show that everything matches except for the `qualifiedName`.

brujoand replied:
I pulled qualifiedName into a method, so going back to quals which makes more sense anyway.

duftler commented on rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/CloudProviderBakeHandler.groovy:
```diff
@@ -140,7 +141,21 @@ abstract class CloudProviderBakeHandler {
+      PackageNameConverter.buildOsPackageName(selectedOptions.baseImage.packageType, name)
+    }
+
+    if (!osPackages.empty) {
```
Can typically just rely on Groovy truthiness when operating on a list (http://mrhaki.blogspot.com/2009/08/groovy-goodness-tell-groovy-truth.html). So this could read: `if (osPackages) {`

brujoand commented:
I've addressed the comments and added tests for empty packages, a test util trait to up the DRY score and aligned the indentation for the test classes so that they are consistent.

duftler commented on rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/util/ImageNameFactory.groovy:
```diff
@@ -16,27 +16,63 @@
-   *
-   * Returns [imageName, appVersionStr, packagesParameter].
+   * Default implementation of ImageNameFactory considers only package_name, a timestamp, and base_os.
+   * For more fine grained conventions, extend this class, override the required methods and alter your
```
Does this work as described here? Does each `CloudProviderBakeHandler` resolve the correct `@Bean` factory method from that provider's configuration class when autowiring the `ImageNameFactory`?
In any case, I think it would be easier to follow if `CloudProviderBakeHandler` defined an abstract `getImageNameFactory` method, instead of relying on the autowiring for this.

brujoand replied:
Good point. I realise that this is much more in line with the current use of the bakeHandler vs roscoConfig also, so I'm changing it.

duftler commented on rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/util/TestDefaults.groovy:
```diff
@@ -0,0 +1,20 @@
+
+import com.netflix.spinnaker.rosco.api.BakeRequest
+
+trait TestDefaults {
```
Nice.

duftler commented on rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/util/ImageNameFactorySpec.groovy:
```diff
@@ -124,13 +133,17 @@ class DefaultImageNameFactorySpec extends Specification {
     when:
-      def (imageName, appVersionStr, packagesParameter) = imageNameFactory.deriveImageNameAndAppVersion(bakeRequest, selectedOptions)
+      def imageName = imageNameFactory.buildImageName(bakeRequest, osPackages)
+      print(imageName)
```
Extraneous `print()`.

brujoand replied:
Ah forgot about that one. 👍

duftler commented on rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandlerSpec.groovy:
```diff
@@ -116,564 +115,582 @@ class OpenstackBakeHandlerSpec extends Specification {
-"""
-
-    Bake bake = openstackBakeHandler.scrapeCompletedBakeResults(REGION, "123", logsContent)
+      def logsContent = """
```
These multiline strings probably shouldn't be reformatted.

brujoand replied:
oh, yes those indents shouldn't have been made part of the actual string. I'll strip the margins so we can still keep the indentation.

duftler commented on rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/openstack/OpenstackBakeHandlerSpec.groovy:
```diff
@@ -116,564 +115,582 @@ class OpenstackBakeHandlerSpec extends Specification {
-"""
-
-    Bake bake = openstackBakeHandler.scrapeCompletedBakeResults(REGION, "123", logsContent)
+      def logsContent = """
```
These multiline strings shouldn't be reformatted.

duftler commented on rosco-core/src/test/groovy/com/netflix/spinnaker/rosco/providers/google/GCEBakeHandlerSpec.groovy:
```diff
@@ -420,13 +440,15 @@ class GCEBakeHandlerSpec extends Specification {
       def imageNameFactoryMock = Mock(ImageNameFactory)
       def packerCommandFactoryMock = Mock(PackerCommandFactory)
       def fullyQualifiedPackageName = "nflx-djangobase-enhanced_0.1-h12.170cdbd_all"
+      def unqualifiedPackageName = "nflx-djangobase-enhanced"
```
Is this intentionally still `unqualified`? Here and in several other tests?

brujoand replied:
Nope, I ment to revert those; doing it now.

duftler commented:
This is looking really nice. LGTM once the final comments are addressed.

brujoand commented:
I've got back in sync with master and addressed the rest of the comments. I kept new indentation on the multiline strings but strip it so we can have both readability and reliability.  Let me know if I should just leave it as it is in master. 

Also I'm pretty sure there are more correct ways of handling the ``` getImageNameFactory ``` methods, but with a private field and a getter in the BakeHandlers I could keep the ease of mocking the ImageNameFactory and still avoid relying on autowireing.

duftler commented:
LGTM.
I agree that the handling of `getImageNameFactory` could be better, but I'll see if I can come up with something after merging this.
